packages/@headlessui-react/src/internal/frozen.tsx

```diff
@@ -0,0 +1,19 @@
+import React, { useState } from 'react'
+
+export function Frozen({ children, freeze }: { children: React.ReactNode; freeze: boolean }) {
+  let contents = useFrozenData(freeze, children)
+  return <>{contents}</>
+}
+
+export function useFrozenData<T>(freeze: boolean, data: T) {
+  let [frozenValue, setFrozenValue] = useState(data)
+
+  // We should keep updating the frozen value, as long as we shouldn't freeze
+  // the value yet. The moment we should freeze the value we stop updating it
+  // which allows us to reference the "previous" (thus frozen) value.
+  if (!freeze && frozenValue !== data) {
+    setFrozenValue(data)
+  }
```

Comment:
This is nice because we don't use any effects, but it will trigger a `setFrozenValue` every time the `data` changes. Ideally we _only_ change this state the moment we go from `false` to `true` and capture the old value somehow. 🤔